src/shared/metrics.c

```diff
@@ -0,0 +1,295 @@
+        assert(link);
+        assert(parameters);
+
+        r = sd_varlink_dispatch(link, parameters, /* dispatch_table= */ NULL, /* userdata= */ NULL);
+        if (r != 0)
+                return r;
+
+        if (!FLAGS_SET(flags, SD_VARLINK_METHOD_MORE))
+                return sd_varlink_error(link, SD_VARLINK_ERROR_EXPECTED_MORE, NULL);
+
+        const MetricFamily *previous = NULL;
+        for (const MetricFamily *mf = metric_family_table; mf && mf->name; mf++) {
+                if (previous) {
+                        r = metric_family_build_send(link, previous, /* more= */ true);
+                        if (r < 0)
+                                return log_debug_errno(r, "Failed to describe metric family '%s': %m", previous->name);
+                }
+
+                previous = mf;
+        }
+
+        if (previous) {
+                r = metric_family_build_send(link, previous, /* more= */ false);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to describe metric family '%s': %m", previous->name);
+        } else
+                return sd_varlink_error(link, "io.systemd.Metrics.NoSuchMetric", NULL);
+
+        return 0;
+}
+
+int metrics_method_list(
+                const MetricFamily metric_family_table[],
+                sd_varlink *link, sd_json_variant *parameters,
+                sd_varlink_method_flags_t flags, void *userdata
+) {
+
+        int r;
+
+        assert(metric_family_table);
+        assert(link);
+        assert(parameters);
+
+        r = sd_varlink_dispatch(link, parameters, /* dispatch_table= */ NULL, /* userdata= */ NULL);
+        if (r != 0)
+                return r;
+
+        if (!FLAGS_SET(flags, SD_VARLINK_METHOD_MORE))
+                return sd_varlink_error(link, SD_VARLINK_ERROR_EXPECTED_MORE, NULL);
+
+        _cleanup_(metric_family_context_done) MetricFamilyContext ctx = {
+                .link = link
+        };
+        for (const MetricFamily *mf = metric_family_table; mf && mf->name; mf++) {
+                assert(mf->generate_cb);
+
+                ctx.metric_family = mf;
+                r = mf->generate_cb(&ctx, userdata);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to list metrics for metric family '%s': %m", mf->name);
+        }
+
+        if (!ctx.previous)
+                return sd_varlink_error(link, "io.systemd.Metrics.NoSuchMetric", NULL);
+
+        /* produce the last metric */
+        return sd_varlink_reply(link, ctx.previous);
+}
+
+static int metric_set_fields(sd_json_variant **v, char **field_pairs) {
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *w = NULL;
+        size_t n_pairs = 0;
+        int r;
+
+        assert(v);
+
+        if (!field_pairs)
+                return 0;
+
+        n_pairs = strv_length(field_pairs);
+        if (n_pairs % 2 != 0)
+                return log_debug_errno(SYNTHETIC_ERRNO(ERANGE), "Odd number of field pairs: %zu", n_pairs);
```

Comment:
A follow up for previous comment. Since you're counting lengths anyway you can just check the result. If it's 0 you cover both NULL and empty array cases.

Also, use `n`. It's a common variable for amount of elements. `n_pairs` is a misleading name anyway because `strv_length()` gives you total amount of elements in strv. Not amount of pairs.

Look into `sd_json_variant_new_object()`  and `sd_json_variant_new_array_strv()` for inspiration.